src/heuristics/cro.rs

```diff
@@ -0,0 +1,176 @@
+/// Basic building blocks of an Chemical Reaction Optimization.
+pub struct Parameters<P> {
+    pub mole_coll: f64,
+    pub kinetic_energy_loss_rate: f64,
+    pub initial_kinetic_energy: f64,
+    pub buffer: f64,
+    pub single_mole_selection: Box<dyn Component<P>>,
+    pub decomposition_criterion: Box<dyn Condition<P>>,
+    pub decomposition: Box<dyn Component<P>>,
+    pub on_wall_ineffective_collision: Box<dyn Component<P>>,
+    pub double_mole_selection: Box<dyn Component<P>>,
+    pub synthesis_criterion: Box<dyn Condition<P>>,
+    pub synthesis: Box<dyn Component<P>>,
+    pub intermolecular_ineffective_collision: Box<dyn Component<P>>,
+}
+
+/// A generic single-objective Chemical Reaction Optimization template.
+///
+/// # References
+/// [doi.org/10.1007/s12293-012-0075-1](https://doi.org/10.1007/s12293-012-0075-1)
+pub fn cro<P: SingleObjectiveProblem>(
+    params: Parameters<P>,
+    termination: Box<dyn Condition<P>>,
+    logger: Box<dyn Component<P>>,
+) -> Box<dyn Component<P>> {
+    let Parameters {
+        mole_coll,
+        kinetic_energy_loss_rate,
+        initial_kinetic_energy,
+        buffer,
+        single_mole_selection,
+        decomposition_criterion,
+        decomposition,
+        on_wall_ineffective_collision,
+        double_mole_selection,
+        synthesis_criterion,
+        synthesis,
+        intermolecular_ineffective_collision,
+    } = params;
+
+    Configuration::builder()
+        .do_(state::CroState::initializer(initial_kinetic_energy, buffer))
+        .while_(termination, |builder| {
+            builder
+                .if_else_(branching::RandomChance::new(mole_coll) | branching::LessThanNIndividuals::new(2),
+                 |builder| {
+                            builder
+                                .do_(single_mole_selection)
+                                .do_(selection::All::new())
+                                .if_else_(decomposition_criterion,
+                             |builder| {
+                                        builder
+                                            .do_(decomposition)
+                                            .evaluate_sequential()
+                                            .update_best_individual()
+                                            .do_(state::CroState::decomposition_update())
+                            },
+                            |builder| {
+                                        builder
+                                            .do_(on_wall_ineffective_collision)
+                                            .evaluate_sequential()
+                                            .update_best_individual()
+                                            .do_(state::CroState::on_wall_ineffective_collision_update(
+                                                kinetic_energy_loss_rate,
+                                            ))
+                            },
+                        )
+                    },
```

Comment:
Why is the formatting so weird?

Reply:
`rustfmt` actually straight up surrenders on this bit and accepts whatever you throw at it. This right there was my attempt of finding a somewhat readable formatting, but seems like the formatting in CLion looks quite different from the one in plain text. I pushed a new formatting, that is hopefully better, but it still looks weird. 

`rustfmt` also occasionally threw an error, even if only whitespace was changed and `clippy` was still happy. 
All very strange, in summary.